### ISSUE TITLE
TensorFlow optimizers / more generally on optimizers

### ISSUE THREAD

**Issue by mikkokotila** [autonomio/talos#111]
It seems that there is a notable speedup just from using `keras.optimizers.TFOptimizer()` and wrapping one of the optimizers in `tf.train` inside it. So for example what would otherwise be: 

```
from keras.optimizers import Adam

...optimizers=Adam
```

Would now become: 

```
from keras.optimizers import TFOptimizer
from tf.train import AdamOptimizer

...optimizers=TFOptimizer(AdamOptimizer)

```

I spent the whole day yesterday working on integrating this in a way where one could just use a string value and then have a switch `tf_optimizer` in Scan() arguments to opt for using a tf optimizer instead of Keras (for the sake of speedup) but there are several important things that make this not so simple. 

First is the use cases: 

1) user inputs string
2) user inputs a Keras optimizer object 
3) user inputs a Tensorflow optimizer object (same as Keras)
4) user inputs a Tensorflow optimizer object (different from Keras)
5) user inputs a custom TensorFlow optimizer object 

The distinction between 3 and 4 is how we already have the learning rate normalizer in place for case 2 (it should also support case 1 or throw a clear error), so case 3 can be handled together with that, whereas case 4 can't. So case 4 and case 5 are similar in that way. Further, it seems that the learning rate normalizer could also be made part of this and have it as an argument in Scan. 

In summary:

| case        | type           | lr_normalizer  |
| ------------- |:-------------:| -----:|
| 1 | str | yes |
| 2 | object | yes |
| 3 | object | yes|
| 4 | object | no |
| 5 | object | no |

This leads to the observation that it's probably better to always first convert string value to the corresponding object, and throw an error unless its one of the 7 available in Keras (more can be added as they might become added to Keras).

Then is the topic of learning rate normalizer, here we have the cases: 

- user inputs actual learning rates (no normalization)
- user inputs no learning rates (something is needed by TensorFlow optimizers) 
- user inputs relative learning rates
   - normalized the current way
   - normalized behind the scenes through Scan() argument `lr_normalize`

So cases 4 and 5 are simple in the sense that they are always handled in the same way, and the optimizer object can be "packaged" in the Talos codes and the user only has to add the optimizers to the params dictionary as they would otherwise. Then Talos simply checks if the inputs are on the list of the 7 optimizers (points 2 and 3 above), and if not, treats them as either standard or custom tf optimizers. 

For the case 1, it will be first converted into object, so now 1 and 2 becomes identical and if `tf_optimizer=True` then as the next step the corresponding tf optimizer is picked, and the packaging is same as with 4 and 5 except if `lr_optimizer=True` where this will be calculated and added (which avoids the need for using lr_normalizer() in the way it's being used now as part of the input model
